facebook/draft-js
# Make aria-describedby able to reference both the placeholder and the description

**Summary**

I've read https://github.com/facebook/draft-js/commit/a6af3e15120e74c8797c5670f5bb63cb45c49a32 and the related issue https://github.com/facebook/draft-js/issues/1214

I'd like to propose an improvement to the way the `placeholder` and `ariaDescribedBy` props work. For more details, please see the related issue #1739.

This PR introduces a new `_getAriaDescribedByIDs()` function to calculate a proper value for the `aria-describedby` attribute.
- it returns `null` when `placeholder` and `ariaDescribedBy` aren't set, to avoid to render an `aria-describedby` attribute that points to nothing 
- when either `placeholder` or `ariaDescribedBy` is set, it just uses its value 
- when both `placeholder` and `ariaDescribedBy` are set, it uses both values, separated by a space 

The third point would allow to replicate the native behavior: when both a `placeholder` and an `aria-describedby` pointing to some description are set in a textarea, assistive technologies use both. 

If you want, you can verify this on a pen I've prepared: https://codepen.io/afercia/full/GdZOmX/

Screen readers announce first the placeholder, then the description referenced by `aria-describedby`. This is the expected behavior, since the `placeholder` and `aria-describedby` have different purposes and `aria-describedby` is not an "override" of the placeholder as mentioned in https://github.com/facebook/draft-js/issues/1214#issuecomment-316835817
Recommended combos to test:
- Safari+VoiceOver
- Firefox+NVDA
- IE11+JAWS

Specifically, the placeholder is meant to help users with data entry, for example providing [a sample value or brief description of the expected format](https://www.w3.org/TR/html52/sec-forms.html#the-placeholder-attribute) while `aria-describedby` is meant for a longer description.

Seems to me allowing both is essential to replicate the native behavior.

Worth noting that currently, when the prop `ariaDescribedBy` is set, the placeholder is not announced at all. As per backwards compatibility concerns, this change shouldn't change so much: everything should work as before, with the added bonus that both props will work.

**Test Plan**

I've tested using the "plaintext" example:
- remove the `placeholder` prop and verify no `aria-describedby` attribute is rendered
- restore the `placeholder` prop and verify the `aria-describedby` attribute value is the same of the placeholder ID
- add an `ariaDescribedBy` with a test ID (you can also add a paragraph with that ID and some content) and verify the `aria-describedby` attribute value takes both IDs
- play with the two props, remove one of them, re-add, etc. and verify the `aria-describedby` attribute value is the expected one


Worth noting `npm run lint` fails for me (also on master, see error below) and `npm run flow` shows a few errors, also on master. Seems to me they're unrelated errors, unless I'm missing something. In that case I'd appreciate any feedback and guidance 🙂 

`npm run lint` error:
```
Cannot read property 'type' of undefined
TypeError: Cannot read property 'type' of undefined
    at hasTypeOfOperator (/Users/andrea/Projects/draft-js/node_modules/eslint/lib/rules/no-undef.js:19:19)
    at Program:exit.globalScope.through.forEach.ref (/Users/andrea/Projects/draft-js/node_modules/eslint/lib/rules/no-undef.js:59:44)
    at Array.forEach (<anonymous>)
    at Program:exit (/Users/andrea/Projects/draft-js/node_modules/eslint/lib/rules/no-undef.js:56:37)
    at listeners.(anonymous function).forEach.listener (/Users/andrea/Projects/draft-js/node_modules/eslint/lib/util/safe-emitter.js:47:58)
    at Array.forEach (<anonymous>)
    at Object.emit (/Users/andrea/Projects/draft-js/node_modules/eslint/lib/util/safe-emitter.js:47:38)
    at NodeEventGenerator.applySelector (/Users/andrea/Projects/draft-js/node_modules/eslint/lib/util/node-event-generator.js:251:26)
    at NodeEventGenerator.applySelectors (/Users/andrea/Projects/draft-js/node_modules/eslint/lib/util/node-event-generator.js:280:22)
    at NodeEventGenerator.leaveNode (/Users/andrea/Projects/draft-js/node_modules/eslint/lib/util/node-event-generator.js:303:14)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! draft-js@0.10.5 lint: `eslint .`
npm ERR! Exit status 1
```
